Move the implementation of register_fake onto torch.library.Library

zou3519 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #124299
* #124200
* #124199
* #124134
* #124180
* #124092
* #124089
* #124071
* #124066
* __->__ #124065
* #124064
* #123937

Motivations:
- This makes things more consistent: using a Library object, you should
  be able to do all of the registration APIs that tie registrations to
  the lifetime of the Library.
- I need this for the next PR up in the stack, where we will have
  torch.library.register_fake support both CustomOpDef (from the new
  custom ops API) and other custom ops.

Test Plan:
- existing tests

zou3519 committed:
Move the implementation of register_fake onto torch.library.Library

Motivations:
- This makes things more consistent: using a Library object, you should
  be able to do all of the registration APIs that tie registrations to
  the lifetime of the Library.
- I need this for the next PR up in the stack, where we will have
  torch.library.register_fake support both CustomOpDef (from the new
  custom ops API) and other custom ops.

Test Plan:
- existing tests

[ghstack-poisoned]

zou3519 committed:
Update on "Move the implementation of register_fake onto torch.library.Library"


Motivations:
- This makes things more consistent: using a Library object, you should
  be able to do all of the registration APIs that tie registrations to
  the lifetime of the Library.
- I need this for the next PR up in the stack, where we will have
  torch.library.register_fake support both CustomOpDef (from the new
  custom ops API) and other custom ops.

Test Plan:
- existing tests

[ghstack-poisoned]

zou3519 committed:
Update on "Move the implementation of register_fake onto torch.library.Library"


Motivations:
- This makes things more consistent: using a Library object, you should
  be able to do all of the registration APIs that tie registrations to
  the lifetime of the Library.
- I need this for the next PR up in the stack, where we will have
  torch.library.register_fake support both CustomOpDef (from the new
  custom ops API) and other custom ops.

Test Plan:
- existing tests

[ghstack-poisoned]

zou3519 committed:
Update on "Move the implementation of register_fake onto torch.library.Library"


Motivations:
- This makes things more consistent: using a Library object, you should
  be able to do all of the registration APIs that tie registrations to
  the lifetime of the Library.
- I need this for the next PR up in the stack, where we will have
  torch.library.register_fake support both CustomOpDef (from the new
  custom ops API) and other custom ops.

Test Plan:
- existing tests

[ghstack-poisoned]

zou3519 committed:
Update on "Move the implementation of register_fake onto torch.library.Library"


Motivations:
- This makes things more consistent: using a Library object, you should
  be able to do all of the registration APIs that tie registrations to
  the lifetime of the Library.
- I need this for the next PR up in the stack, where we will have
  torch.library.register_fake support both CustomOpDef (from the new
  custom ops API) and other custom ops.

Test Plan:
- existing tests

[ghstack-poisoned]

zou3519 committed:
Update on "Move the implementation of register_fake onto torch.library.Library"


Motivations:
- This makes things more consistent: using a Library object, you should
  be able to do all of the registration APIs that tie registrations to
  the lifetime of the Library.
- I need this for the next PR up in the stack, where we will have
  torch.library.register_fake support both CustomOpDef (from the new
  custom ops API) and other custom ops.

Test Plan:
- existing tests

[ghstack-poisoned]

zou3519 committed:
Update on "Move the implementation of register_fake onto torch.library.Library"


Motivations:
- This makes things more consistent: using a Library object, you should
  be able to do all of the registration APIs that tie registrations to
  the lifetime of the Library.
- I need this for the next PR up in the stack, where we will have
  torch.library.register_fake support both CustomOpDef (from the new
  custom ops API) and other custom ops.

Test Plan:
- existing tests

[ghstack-poisoned]

albanD commented on torch/library.py:
```diff
@@ -125,6 +125,30 @@ def define(self, schema, alias_analysis="", *, tags=()):
+        # (e.g. __main__)
+        caller_module_name = None if caller_module is None else caller_module.__name__
+
+        # TODO(rzou): We're gonna need to stage this change with torchvision,
```
Not sure what this means?
Is it just: "TODO: update torchvision and remove this code that is here just for BC"

zou3519 replied:
Yes. I didn't add the comment; it was moved from its original location

albanD commented on torch/library.py:
```diff
@@ -491,34 +515,23 @@ def register_fake(qualname, func=None, /, *, lib=None, _stacklevel=1):
-            func_to_register = _check_pystubs_once(func, qualname, caller_module_name)
-        else:
-            func_to_register = func
+    stacklevel = _stacklevel
```
Why this rename here and not in the arg?
Private args in python are a lie anyways, everyone can see them haha

zou3519 replied:
I needed to update the stacklevel below and didn't want to modify the `_stacklevel` argument

albanD reviewed (approved):
Bit more tricky than I was expecting, but sounds fair.

zou3519 committed:
Update on "Move the implementation of register_fake onto torch.library.Library"


Motivations:
- This makes things more consistent: using a Library object, you should
  be able to do all of the registration APIs that tie registrations to
  the lifetime of the Library.
- I need this for the next PR up in the stack, where we will have
  torch.library.register_fake support both CustomOpDef (from the new
  custom ops API) and other custom ops.

Test Plan:
- existing tests

[ghstack-poisoned]

zou3519 commented:
@pytorchbot merge
